### ISSUE TITLE
command_not_found hook - add a way to "not fail"

### ISSUE THREAD

**Issue by AsafMah** [nushell/nushell#12468]
### Related problem

I want to implement an automatic cd with zoxide - if the command name doesn't exist, it will instead check if it's a directory and try to navigate there.

However, when using the command_not_found there are two problems:  
1. No matter what I return (string or null), it still ends up being an error. 
2. Environment changes, such as the `cd` command, aren't saved.

### Describe the solution you'd like

1. Have some way for command_not_found to either return a value that means "I have handled this", or return a command that will be ran instead of the invalid one.
2. Keep env changes within the block.

**Comment by fdncred** [comment]
There are instructions on the zoxide repo of how to properly implement it in nushell. I've been using it for years and it works well. I'm not really sure why you'd need to use the command_not_found hook at all for zoxide.

**Comment by AsafMah** [comment]
I'll be clearer: Zoxide works fine currently, but I want to extend its abilities.

Currently, let's say I want to navigate to the directory `repos`, I need to type `z repos`, that's the current functionality.

What I want to do, is simply type `repos`, and since there isn't a command or a file with that name, it will act the same as doing `z repos`.

I haven't looked at the implementation until now, just the doc, but looking at it now confirms what I said - the hook can only control the error message being returned, but not actually prevent the error and replace it with a different command.

Other than this specific use case, I can think of some others - autocorrect of spelling mistakes, ad-hoc syntax for special commands, etc.

**Comment by fdncred** [comment]
Agreed. Looking at the examples, it's built for 1 intended use. We'd probably be open to extending it for a use case like you describe so it supports both use cases.

**Comment by pfmoore** [comment]
I would also find this useful. My use case would be for having a fallback for when a command is not found - for example, to attempt to run `uutils <cmd>` if the command wasn't found (where uutils is the Rust reimplementation of GNU coreutils - I don't want to install all of the individual commands as some of them clash with "native" Windows commands like sort which might be needed in `.bat` files).

**Comment by eifr** [comment]
Facing the same problem. I'm using [pkgx](https://github.com/pkgxdev/pkgx/) and they have an option to try to use their bins in case of not found in path (see: https://github.com/pkgxdev/pkgx?tab=readme-ov-file#magic) 
this option is problematic atm as it cuts off outputs if I'm using the hook to achieve this feature.

![Image](https://github.com/user-attachments/assets/330dbafc-f705-42d0-b670-2b2ce3ccc7f4)

**Comment by youwen5** [comment]
I have the same use case, with Nix. In `fish`, if I type a command not on my machine, it gets handled by the NixOS command not found hook:
```
$ asciiquarium                                                               ⌂ 16:04

The program 'asciiquarium' is currently not installed. It is provided by
several packages. You can install it by typing one of the following:
  nix profile install nixpkgs#asciiquarium-transparent.out
  nix profile install nixpkgs#asciiquarium.out

Or run it once with:
  nix shell nixpkgs#asciiquarium-transparent.out -c asciiquarium ...
  nix shell nixpkgs#asciiquarium.out -c asciiquarium ...
```
I'd like a way to hook into this in Nushell.

**Comment by dreamgineer** [comment]
My current setup looks like this:
```nu
# pay-respects
def --env f [] {
	let dir = (with-env { _PR_LAST_COMMAND: (history | last).command, _PR_ALIAS: (help aliases | select name expansion | each ({ |row| $row.name + "=" + $row.expansion }) | str join (char nl)), _PR_SHELL: nu } { ~/.nix-profile/bin/pay-respects })
	cd $dir
}

$env.config.hooks.command_not_found = {
    |command: string|
	let dir = (with-env { _PR_LAST_COMMAND: $command, _PR_ALIAS: (help aliases | select name expansion | each ({ |row| $row.name + "=" + $row.expansion }) | str join (char nl)), _PR_SHELL: nu } { ~/.nix-profile/bin/pay-respects })
	cd $dir
}
```

I am trying to get rid of Nushell's error. But couldn't.
```
~ fstfetch
1 suggestion(s) found
> 
 fastfetch
          ▗▄▄▄       ▗▄▄▄▄    ▄▄▄▖             dgnr@nyx
          ▜███▙       ▜███▙  ▟███▛             --------
           ▜███▙       ▜███▙▟███▛              OS: NixOS 25.05 (Warbler) x86_64
... normal fastfetch stuff ...                                
Error: nu::shell::external_command

  × External command failed
   ╭─[entry #41:1:1]
 1 │ fstfetch
   · ────┬───
   ·     ╰── Command `fstfetch` not found
   ╰────
  help: `fstfetch` is neither a Nushell built-in or a known external command

~ 
```

**Comment by manveru** [comment]
The way I worked around this is:
```
$env.config.hooks.command_not_found = {|cmd: string|
  hide-env ENTERED_COMMAND_NOT_FOUND
  exec nu -e $", ($cmd)"
}
```
My remaining problem is that it will ignore all arguments, and I haven't found a way to make it take more than one argument.
Also hiding `ENTERED_COMMAND_NOT_FOUND` may summon Shoggoths or sacrifice your first-born, so handle with care.
